src/main/java/seedu/address/model/ModelManager.java

```diff
@@ -86,36 +94,50 @@ public void setAgentAssistFilePath(Path agentAssistFilePath) {
 
     @Override
     public void setAgentAssist(ReadOnlyAgentAssist agentAssist) {
-        this.agentAssist.resetData(agentAssist);
+        safeHistory();
+        this.currentAgentAssist.resetData(agentAssist);
     }
 
     @Override
     public ReadOnlyAgentAssist getAgentAssist() {
-        return agentAssist;
+        return currentAgentAssist;
     }
 
     @Override
     public boolean hasPerson(Person person) {
         requireNonNull(person);
-        return agentAssist.hasPerson(person);
+        return currentAgentAssist.hasPerson(person);
     }
 
     @Override
     public void deletePerson(Person target) {
-        agentAssist.removePerson(target);
+        safeHistory();
+        currentAgentAssist.removePerson(target);
     }
 
     @Override
     public void addPerson(Person person) {
-        agentAssist.addPerson(person);
+        safeHistory();
+        currentAgentAssist.addPerson(person);
         updateFilteredPersonList(PREDICATE_SHOW_ALL_PERSONS);
     }
 
     @Override
     public void setPerson(Person target, Person editedPerson) {
         requireAllNonNull(target, editedPerson);
+        safeHistory();
+        currentAgentAssist.setPerson(target, editedPerson);
+    }
 
-        agentAssist.setPerson(target, editedPerson);
+    @Override
+    public void undoCommand() {
+        requireNonNull(this.historyAgentAssist);
+        this.setAgentAssist(this.historyAgentAssist);
+    }
+
+    @Override
+    public boolean hasPreviousCommand() {
+        return this.historyAgentAssist != null;
     }
```

Comment:
Great use of abstraction! Liked how clean this is with just adding saveHistory into each command works.